kanidmd/src/lib/be/idl_arc_sqlite.rs

```diff
@@ -710,6 +716,257 @@ impl<'a> IdlArcSqliteWriteTransaction<'a> {
+    /// weights. The intent is to seperate over obviously more effective indexes rather than
+    /// to min-max the fine tuning of these. Consider name=foo vs class=*. name=foo will always
+    /// be better than class=*, but comparing name=foo to spn=foo is "much over muchness" since
+    /// both are really fast.
+    pub fn analyse_idx_slopes(&mut self, audit: &mut AuditScope) -> Result<(), OperationError> {
+        /*
+         * Inside of this analysis there are two major factors we need to understand
+         *
+         * * What is the variation of idl lengths within an index?
+         * * How man keys are stored in this index?
+         *
+         * Since we have the filter2idl threshold, we want to find "what is the smallest
+         * and most unique index asap so we can exit faster". This allows us to avoid
+         * loading larger most costly indexs that either have large idls, high variation
+         * or few keys and are likely to miss and have to go out to disk.
+         *
+         * A few methods were proposed, but thanks to advice from Perri Boulton (psychology
+         * researcher with a background in statistics), we were able to device a reasonable
+         * approach.
+         *
+         * These are commented in line to help understand the process.
+         */
+
+        /*
+         * Step 1 - we have an index like "idx_eq_member". It has data that looks somewhat
+         * like:
+         *
+         *  key    | idl
+         *  -------+------------
+         *  uuid_a | [1, 2, 3, ...]
+         *  -------+------------
+         *  uuid_b | [4, 5, 6, ...]
+         *
+         * We need to collect this into a single vec of "how long is each idl". Since we have
+         * each idl in the vec, the length of the vec is also the number of keys in the set.
+         * This yields for us:
+         *
+         *   idx_eq_member: [4.0, 5.0, ...]
+         * where each f64 value is the float representation of the length of idl.
+         *
+         * We then assemble these to a map so we have each idxkey and it's associated list
+         * of idl lens.
+         */
+
+        let mut data: HashMap<IdxKey, Vec<f64>> = HashMap::new();
+        self.idl_cache.iter_dirty().for_each(|(k, maybe_idl)| {
+            if let Some(idl) = maybe_idl {
+                let idl_len: u32 = idl.len().try_into().unwrap_or(u32::MAX);
+                // Convert to something we can use.
+                let idl_len = f64::from(idl_len);
+
+                let kref = IdxKeyRef::new(&k.a, &k.i);
+                if idl_len > 0.0 {
+                    // It's worth looking at. Anything len 0 will be removed.
+                    if let Some(lens) = data.get_mut(&kref as &dyn IdxKeyToRef) {
+                        lens.push(idl_len)
+                    } else {
+                        data.insert(kref.to_key(), vec![idl_len]);
+                    }
```

Comment:
Whoa boy no. The set holds an IdxKey not an IdxKeyRef. And the ref has a lifetime, and that lifetime has to survive multiple transactions. So we have to store IdxKeys. Then if we want to do a lookup we need K: Borrow<Q> where the input is Q. https://docs.rs/hashbrown/0.11.2/hashbrown/struct.HashMap.html#method.get

So either:
* We have to allocate and clone a new IdkKey every time in the loop to do the lookup
* We create a way that a reference can borrow into IdxKey. 

https://github.com/kanidm/kanidm/blob/master/kanidmd/src/lib/be/idxkey.rs#L7

This blog here explains the magci. Both IdxKey and IdxKeyRef impl dyn IdxKeyToRef, and that's the common Borrow implementation that we can use. 

This actually saves *tons* of time in lookups compared to the cloning / allocation strategy. Especially in some of the hotter paths of the code.

Reply:
So in theory, we could've made it take an `IdxKey` and then cloned for each lookup, and then the `get_mut` argument and the `insert` argument would've taken the same type, but instead we're doing this to avoid cloning since it's more expensive, especially since it contains `String`.

Also, I think we can delete this `impl` block because it doesn't seem to be used (?)
https://github.com/kanidm/kanidm/blob/93d5c5434d67ee6231526b9de9f0c6d79f012bb2/kanidmd/src/lib/be/idxkey.rs#L21-L25

Reply:
I think was experimenting with using it, but honestly, it's not harmful to leave it.